### PR TITLE
Fix unsynchronized access in InProcessTransport

### PR DESCRIPTION
**Pull request by carl-mastrangelo**

### CODE REVIEW

File: core/src/main/java/io/grpc/inprocess/InProcessTransport.java [hunk 1]
```diff
@@ -123,7 +123,9 @@ public synchronized ClientStream newStream(
       return new NoopClientStream() {
         @Override
         public void start(ClientStreamListener listener) {
-          listener.closed(shutdownStatus, new Metadata());
+          synchronized (InProcessTransport.this) {
+            listener.closed(shutdownStatus, new Metadata());
```
Comment (nmittler) [inline core/src/main/java/io/grpc/inprocess/InProcessTransport.java]: why is this needed? How would the listener be called asynchronously if we're only just starting it now?

Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/inprocess/InProcessTransport.java]: The GaurdedBy annotation causes a build failure internally without this.   It's actually the shutdownStatus that has the problem.

Comment (ejona86) [inline core/src/main/java/io/grpc/inprocess/InProcessTransport.java]: Nowhere else in this file are we synchronizing on `InProcessTransport` before calling `listener` (we do synchronize on the streams, though), and that doesn't seem like something we want to start doing.

Comment (ejona86) [inline core/src/main/java/io/grpc/inprocess/InProcessTransport.java]: Ah, I see. Go ahead and save the shutdownStatus within newStream as a final local variable, then.

Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/inprocess/InProcessTransport.java]: Done.

**Comment by ejona86** [comment]
LGTM

**Comment by nmittler** [comment]
LGTM
